lapack/testlapack/dgetc2.go

```diff
@@ -13,86 +13,123 @@ import (
+		for j := i + 1; j < n; j++ {
+			l.Data[i*l.Stride+j] = 0
+		}
 	}
-	// Generate a random general matrix A.
-	a := randomGeneral(n, n, lda, rnd)
-	// ipiv and jpiv are outputs.
+
+	// Generate a random upper-triangular matrix.
+	u := randomGeneral(n, n, max(1, n), rnd)
+	for i := 0; i < n; i++ {
+		for j := 0; j < i; j++ {
+			u.Data[i*u.Stride+j] = 0
+		}
+	}
+	if perturb && n > 0 {
+		// Make U singular by randomly placing a zero on the diagonal.
+		i := rnd.Intn(n)
+		u.Data[i*u.Stride+i] = 0
+	}
+
+	// Construct A = L*U.
+	a := zeros(n, n, max(1, lda))
+	blas64.Gemm(blas.NoTrans, blas.NoTrans, 1, l, u, 0, a)
+
+	// Allocate slices for pivots and pre-fill them with invalid indices.
 	ipiv := make([]int, n)
 	jpiv := make([]int, n)
 	for i := 0; i < n; i++ {
-		ipiv[i], jpiv[i] = -1, -1 // Set to non-indices.
+		ipiv[i] = -1
+		jpiv[i] = -1
 	}
-	// Copy to store output (LU decomposition).
+
+	// Call Dgetc2 to compute the LU decomposition.
 	lu := cloneGeneral(a)
 	k := impl.Dgetc2(n, lu.Data, lu.Stride, ipiv, jpiv)
-	if k >= 0 {
-		t.Logf("%v: matrix was perturbed at %d", name, k)
+
+	if n == 0 {
+		return
 	}
 
-	// Verify all indices are set.
+	if perturb && k < 0 {
+		t.Errorf("%v: expected matrix perturbation", name)
+	}
+
+	// Verify all indices have been set.
 	for i := 0; i < n; i++ {
 		if ipiv[i] < 0 {
-			t.Errorf("%v: ipiv[%d] is negative", name, i)
+			t.Errorf("%v: ipiv[%d] is not set", name, i)
 		}
 		if jpiv[i] < 0 {
-			t.Errorf("%v: jpiv[%d] is negative", name, i)
+			t.Errorf("%v: jpiv[%d] is not set", name, i)
 		}
 	}
-	bi := blas64.Implementation()
-	// Construct L and U triangular matrices from Dgetc2 output.
-	L := zeros(n, n, lda)
-	U := zeros(n, n, lda)
+
+	// Construct L and U matrices from Dgetc2 output.
+	l = zeros(n, n, n)
+	u = zeros(n, n, n)
 	for i := 0; i < n; i++ {
-		for j := 0; j < n; j++ {
-			idx := i*lda + j
-			if j >= i { // On upper triangle and setting of L's unit diagonal elements.
-				U.Data[idx] = lu.Data[idx]
-				if j == i {
-					L.Data[idx] = 1.0
-				}
-			} else if i > j { // On diagonal or lower triangle.
-				L.Data[idx] = lu.Data[idx]
-			}
+		for j := 0; j < i; j++ {
+			l.Data[i*l.Stride+j] = lu.Data[i*lu.Stride+j]
+		}
+		l.Data[i*l.Stride+i] = 1
+		for j := i; j < n; j++ {
+			u.Data[i*u.Stride+j] = lu.Data[i*lu.Stride+j]
 		}
 	}
-	work := zeros(n, n, lda)
-	bi.Dgemm(blas.NoTrans, blas.NoTrans, n, n, n, 1, L.Data, L.Stride, U.Data, U.Stride, 0, work.Data, work.Stride)
+	diff := zeros(n, n, n)
+	blas64.Gemm(blas.NoTrans, blas.NoTrans, 1, l, u, 0, diff)
 
-	// Apply Permutations P and Q to L*U.
+	// Apply permutation matrices P and Q to L*U.
 	for i := n - 1; i >= 0; i-- {
-		ipv, jpv := ipiv[i], jpiv[i]
+		ipv := ipiv[i]
 		if ipv != i {
-			bi.Dswap(n, work.Data[i*lda:], 1, work.Data[ipv*lda:], 1)
+			row1 := blas64.Vector{N: n, Data: diff.Data[i*diff.Stride:], Inc: 1}
+			row2 := blas64.Vector{N: n, Data: diff.Data[ipv*diff.Stride:], Inc: 1}
+			blas64.Swap(row1, row2)
```

Comment:
I'm not a fan of using Swap over Dswap.

Reply:
Yeah, it's a bit verbose but on the other hand it forces me to have `row1`, `col1`, ... variables which makes the code more explicit and self-documenting. And all the BLAS calls are made through `blas64` package level functions.